---
title:  "Building from Linear Models to Generalised Linear Models: Part 2"
author: "Emma Rand"
output:
  html_document:
    toc: true
    depth: 3
    toc_float:
      collapsed: false
      smooth_scroll: false
    theme: yeti
  word_document: default
---
![](../../pics/58I.png)

  

# Introduction

## Aims
In the second of two related workshops we will learn to apply and interpret the glm() function.

## Objectives 
By carrying out the Prior independent learning, working through workbook examples during the workshop workshop and completing follow-up independent study the successful student will be able to:

* Explain the link between the general linear models and the generalised linear model
* Recognise where a generalised linear model would be appropriate and apply `glm()`
* Determine which effects are significant using using `summary()` and `anova()`

You can optionally stretch yourself by asking for more in-depth explanations about the meaning of the estimates, and the direction and magnitude of the effects or creating figures to go with your analyses. Biomedical Scientists might be particularly interested in binomial glm estimates ('odds ratios').

# Prior Independent Learning

Watch these three videos which introduce the workshop, explain the link between the general linear models and the generalised linear model, binomial and Poisson glms, and how to interpret the output of the `glm()` function.

* [GLM Overview (~ 5:00)](https://york.cloud.panopto.eu/Panopto/Pages/Viewer.aspx?id=c5175b27-89d4-4d5e-9100-a96c0091097a)
* [Poisson GLM (~ 6:00)](https://york.cloud.panopto.eu/Panopto/Pages/Viewer.aspx?id=7338e6a9-e104-49d9-933c-a96c00910897)
* [Binomial GLM (~ 7:30)](https://york.cloud.panopto.eu/Panopto/Pages/Viewer.aspx?id=4219cedc-dc3a-4b3c-b56b-a96c00910800)

The slides accompanying the videos are here: [GLMS](../slides/GLMs.pdf)

Please note, the module number was 32I last year. The slides have been updated to show 58I but the videos show 32I slides. The content is unchanged.


# Workbook

## About the Workbook

The workbook for this session is divided in to 2 sections. 

**You are not expected do all of the workbook examples**

Choose **two** from each section that best matches your biological interests. For each example you choose, you should:

* read in the data file, check you know the structure and can identify the response and explanatory variables
* build a model with glm() using examples on the slides
* examine the model with summary()
* interpret the results

Optional Extension: Practice your plotting skills.


## Section 1

**Choose two of:**

----------------------------------------

### Wolf Spiders

This example concerns the effect of sand grain size on the presence of wolf spiders.
Suzuki et al. (2006) measured sand grain size on 28 beaches in Japan and observed the presence or absence of the burrowing wolf spider _Lycosa ishikariana_ on each beach. The data are in [grainsize.txt](../data/grainsize.txt). Can you predict the presence of spiders from the sand grain size? 

```{r echo=FALSE,results="hide",warning=FALSE,message=FALSE,eval=FALSE}
#============== WORKBOOK EXAMPLE ==============#
# The effect of sand grain size on the presence of wolf spiders
#read in file
grainsize <- read.table("../data/grainsize.txt",header=T)
str(grainsize)
#build and examine model. binary outcome so it is binomial
mod <- glm(spiders ~ gsize,family = binomial, data = grainsize)
summary(mod) 
#the probability of spiders being present increases with sand grainsize
anova(mod, test="Chisq")
#there is a signifcant reduction in deviance when gsize is included as an explanatory variable
exp(mod$coef)
# to understand the results a bit better we can get the predict probabilities of advanced amd using
#probabilites for gsize from 0 to 1 in steps of 0.1
newdat <- data.frame(gsize = seq(0,1,0.1))
newdat
newdat$pr <- predict(mod, type = "response", newdata = newdat)
newdat

#this allows us to see how the probability of spiders being present increases with grain size.
#note the the odds ratio seems huge because it is comparing odds at grain sizes of zero and 1 (i.e., per unit)

#(0.9699368/(1-0.9699368))/(0.1614302/(1-0.1614302)) advanced for ref
# plotting for ref
library(ggplot2)
grainsize$spiders<-as.numeric(grainsize$spiders)-1

ggplot(grainsize, aes(x = gsize, y = spiders))+
  geom_point() +
  stat_smooth(method = "glm", method.args = list(family = "binomial"),
  col="black",se=FALSE)


#important: latest versions of ggplot you need method.args = list(family = "binomial")
#older versions of ggplot just family = "binomial"
```

----------------------------------------

### Parkinson's disease 

This example examines the progression of Parkinson's disease in flies with a Parkinson's disease-associated mutation.
The effect of age on locomoter ability of flies with a Parkinson's disease-associated mutation was determined by using a climbing assay. Ten flies of each tested age were placed in 3 replicate vials and the number able to climb to a set height within set time was recorded.
The data in are [park.txt](../data/park.txt) and the ages are given in days. Each row is a vial. Can you predict the number of mutant flies able to climb by their age?  

```{r echo=FALSE,results="hide",warning=FALSE,message=FALSE,eval=FALSE}
#============== WORKBOOK EXAMPLE ==============#
# Disease progression in Parkinson's disease-associated mutants
# Read in file. the response is counts so poisson is appropriate
park <- read.table("../data/park.txt",header=T)
str(park)
mod<-glm(flies~age, family="poisson", data=park)
summary(mod)
exp(mod$coef)
anova(mod,test="Chisq")
# interpret the results
 #6.9 flies are expected to be able to climb at age 0 (the intercept). For each day they age, the number that can climb is reduced by a factor of 0.96. This is significant
```

----------------------------------------

### Protein kinase

This example concerns the effect of a Mitogen-activated protein (MAP) Kinase Inhibitor on the number of nuclei in neurons
The importance of Mitogen-activated protein kinases in regulating cell division led researchers to hypothesise that MAPK inhibition might effect cytokinesis following mitosis. They treated samples of neurons with PD089059, a MAPK inhibitor, at various concentrations (1 to 15 in arbitrary units) and recorded the number of nuclei per cell.
The data in are [kinase.txt](../data/kinase.txt). Each row is a cell. Can you predict the number of nuclei by PD089059 concentration?  

```{r echo=FALSE,results="hide",warning=FALSE,message=FALSE,eval=FALSE}
#============== WORKBOOK EXAMPLE ==============#
# The effect of a MAPK Kinase Inhibitor on the number of nuclei in neurons
# Read in file. the response is counts so poisson is appropriate
kinase <- read.table("../data/kinase.txt",header=T)
str(kinase)
mod<-glm(nuclei~pdconc, family="poisson", data=kinase)
summary(mod)
exp(mod$coef)
anova(mod,test="Chisq")
# interpret the results
 #1.008 nuclei are expected PDconc = 0 (the intercept). For each unit of concentration increase the number of nuclei increases by a factor of 1.09. This is significant
```

----------------------------------------

### Oesophageal cancer

This examples examines the effect of alcohol consumption on the incidence of oesophageal cancer in men over 55 years of age.
Thirty men aged 55 years and over were survey for their alcohol consumption then followed up 10 years later for the occurrence of oesophageal cancer. The data are in [oesoph.txt](../data/oesoph.txt) and comprise two variables:

* status : a variable which indicates whether the individual had developed oesophageal cancer (1) or not  (0)
* alcohol :  the amount of alcohol consumed per week in grams


```{r echo=FALSE,results="hide",warning=FALSE,message=FALSE,eval=FALSE}
#============== WORKBOOK EXAMPLE ==============#
# The effect of sand grain size on the presence of wolf spiders
#read in file
oesoph <- read.table("../data/oesoph.txt",header=T)
str(oesoph)
#build and examine model. binary outcome so it is binomial
mod <- glm( status ~ alcohol, data = oesoph,family="binomial")
summary(mod)
#the probability of oesophageal cancer increases with alcohol
anova(mod, test="Chisq")
#there is a signifcant reduction in deviance when alcohol is included as an explanatory variable
exp(mod$coef)
# to understand the results a bit better we can get the predicted probabilities oesophageal cancer increases using predict()
#probabilites for alcohol from 0 to 80 in steps of 10
newdat<-data.frame(alcohol=seq(0,80,10))
newdat
newdat$pr<-predict(mod,type="response",newdata=newdat)
newdat
#this allows us to see how the probability increases with alcohol.

#advanced for ref
(0.006795236/(1-0.006795236)) # intercept
(0.015739186/(1-0.015739186)) /(0.006795236/(1-0.006795236)) #slope^10 as odds are fat 0 and 10 grrams and slope is 0 to 1 grams
1.088605775^10
```

## Section 2

**Choose two of:**

----------------------------------------

### Age-related macular degeneration

This example concerns the Progression of moderate age-related macular degeneration to advanced AMD.
Age-related macular degeneration (AMD) is a deterioration the macula, a small area in the retina responsible for central vision. There is no cure and treatment is focused on slowing the progression of moderate AMD to advanced AMD and the resulting vision loss. There is variation in response a treatment called AREDS and in order to understand this, researchers determined a person's genotype,  their refined carbohydrate intake and whether they developed advanced AMD after 7 years of AREDS treatment. The data are in  [amd.txt](../data/amd.txt) and comprise the following variables:

* aamd: whether or not an individual developed advanced AMD after 7 years (0 or 1)
* HRC: a continuous measure of the amount of highly refined carbs (kg per day)
* genotype: a factor with two levels,ARMS2 and CFH

Each row is an individual

```{r echo=FALSE,results="hide",warning=FALSE,message=FALSE,eval=FALSE}
# The Progression of moderate age-related macular degeneration to advanced AMD
# read in file
amd <- read.table("../data/amd.txt",header=T)
str(amd)

#build and examine model. the repsonse is binary so it is binomial
mod <- glm(data=amd,aamd~ HRC*alleles, family=binomial)
summary(mod)
exp(mod$coefficients)
anova(mod, test="Chisq")
#there is a sig efect of allele and sig effect of HRC but not interaction between them (so the effect of HRC is the same for both genotypes). As HRC increased risk of advanced amd increases; allele CFH carries a greater risk.
#since the interaction is not sig we could drop it

#mod2, without interaction
mod2<-glm(data=amd,aamd~ HRC+alleles, family=binomial)
anova(mod, mod2,test="Chisq") #confirms it
summary(mod2)
anova(mod2, test="Chisq")

# to understand the results a bit better we can get the predict probabilities of advanced amd using
#probabilites for each allele at HRC of 0, 0.5 and 1.0
newdat<-data.frame(HRC=rep(seq(0,1,0.5),2),
   alleles=rep(c("ARMS2","CFH"),each=3))
newdat
newdat$pr<-predict(mod2,type="response",newdata=newdat)
newdat
#this allows us to see that the risk of advanced amd if quite a bit higher for CFH

 
```

----------------------------------------

### Number of mutations

This example is about how the number of mutations in a hypermutable tetranucleotide marker is affected by a person's age and whether of not they have cancer.
The number of mutations in the hypermutable tetranucleotide marker D7S1482 were analysed in buccal specimens from 30 head and neck carcinoma cases and 43 controls. Also recorded was the subject's age (in years). The goal of analysis was to determine whether age and cancer status could predict the number of mutations. The data are in  [mutation.txt](../data/mutation.txt) and comprise the following variables:

* mut : the number of mutations
* age : a continuous measure of the subject's age to the nearest 0.1 of a degree.
* cat : a factor with two levels, "control" and "tumour"


```{r echo=FALSE,results="hide",warning=FALSE,message=FALSE}
#============== WORKBOOK EXAMPLE ==============#
#The effect of age and cancer status on the number of mutations in a hypermutable tetranucleotide marker
#Read in the data file 
mutation <- read.table("../data/mutation.txt", header = TRUE)
str(mutation)
#build and examine model. the repsonse is a count so it is poisson
mod<-glm(mut~age*cat, family="poisson", data=mutation)
summary(mod)
anova(mod,test="Chisq")


# There is a significant effect of age on the number of mutations with the number of mutations increasing with age
#The control subjects had signifcantly fewer mutation than tumour subjects 
#the interaction was signifcant: The effect of age is greater for tumour subjects than control subjects 


```

----------------------------------------

### Skin micro-organisms

Human skin is colonised by a diverse collection of micro-organisms which vary considerably between individuals.  The presence or absence of a particular micro-organism on the skin of a number of individuals was determined along with variables which might influence presence. The data are in [microrg.txt](../data/microrg.txt) and comprise following variables:

* melanin : a continuous measure of the concentration of melanin in the individual's skin determined by the SR method
* age     : the individual's age in years (to one tenth of a year) 
* presence: whether the micro-organism is absent (0) or present (1) on the individual's skin
* gender  : female or male

The goal of analysis was to determine if the presence of the micro-organism could be predicted from an individual's gender, melanin concentration or age.

```{r echo=FALSE,results="hide",warning=FALSE,message=FALSE,eval=FALSE}
#============== WORKBOOK EXAMPLE ==============#
#Determinants of skin micro-organism presence
#read in data
micro <- read.table("../data/microrg.txt",header=T)
str(micro)
#build and examine model. the repsonse is binary so it is binomial
   mod <- glm(data=micro,presence ~ gender*melanin*age, family=binomial)
   summary(mod)
   anova(mod,test="Chisq")
# interpret the results
# only gender and melanin are signifcant: the presence of microorganisms seems to be associated with lower melanin and male gender.
   
# More advanced: to understand the results a bit better we can get the predicted probabilities of presence for particular melanin and gender
summary(micro$melanin)   
newdat<-data.frame(melanin=rep(seq(0,12,2),2),
           gender=rep(c("male","female"),each=7),
                   age=mean(micro$age))
newdat
newdat$pr<-predict(mod,type="response",newdata=newdat)
newdat
#this allows us to see that the probabilities of presence are higher for low melanin at average age
   #in reality we would likely simplify the model starting with removing the three-way interaction
```

----------------------------------------

### Birds catching insects

This example considers the effects of age, habitat type and group size on the ability of birds to catch insects.
The number of insect prey individuals of a particular bird species manages to collect varies. In an effort to understand this variation, researchers recorded the number of prey an individual caught, its age, whether it was foraging in open or closed habitat and how it spent the majority of it's time (as a single individual, in a pair or in a group of many). The data are in [prey.txt](../data/prey.txt) and comprise following variables:

* prey    : the number of insect prey items caught
* age     : the individual's age in years (to one tenth of a year) 
* group   : how the individual spent the majority of its time; a factor with three levels "many", "pair", "single"
* habitat : a factor with 2 levels "closed", "open"

The goal of analysis was to determine if the number of prey items caught could be explained by age, habitat and group.

```{r echo=FALSE,results="hide",warning=FALSE,message=FALSE,eval=FALSE}
#============== WORKBOOK EXAMPLE ==============#
#The effect of age, habitat and group size on the ability of birds to catch insects
#read in data
prey <- read.table("../data/prey.txt",header=T)
str(prey)
#build and examine model. the repsonse is counts so it is poisson
   mod <- glm(prey ~ age * group * habitat, family = poisson, data=prey)
   summary(mod)
   anova(mod,test="Chisq")
# interpret the results
# there is a signifcant effect of age with old birds catching more insects. there is a signifcant effect of group size with individuals spend time with many others catching the most insects and singles catching the fewest. there is also an interaction between group size and age. it is hard to tease out in what waht way from the summary table but it looks like the effect of age is smaller for singles.
   #in reality we would likely simplify the model starting with removing the three-way interaction 
```

# The Rmd file

Suggested analyses and interpretation for Workbook examples are marked:

`#============== WORKBOOK EXAMPLE ==============#` 


[Rmd file](QCworkshop3DataAnalysis.Rmd)


# Objectives from previous sessions

[Thinking about data generation and processing before experimental design and analysis](QCworkshop1DataAnalysis.html)

* Recognise non-independence and pseudo replication in experimental design
* Select appropriately, and apply some methods to make data comparable
* Design experiments to take account of these

[Building from Linear Models to Generalised Linear Models Part 1](QCworkshop2DataAnalysis.html)

* Explain the the link between t-tests, ANOVA and regression
* Appropriately apply linear models using `lm()`
* Interpret the results using `summary()` and `anova()` and relate them to the outputs of `t.test()` and `aov()`

[Building from Linear Models to Generalised Linear Models: Part 2](QCworkshop3DataAnalysis.html)

* Explain the link between the general linear models and the generalised linear model
* Recognise where a generalised linear model would be appropriate and apply `glm()`
* Determine which effects are significant using using `summary()` and `anova()`

# Q and C Overview

[Quantitative and Computational skills Intro.pdf](../slides/Quantitative and Computational skills Intro.pdf)

This series of workshops is designed to allow you to consolidate and extend your quantitative and computational skills. We build on the experimental design and data analysis skills developed in Stage 1 by introducing you to additional methods of data analysis as well as those for data handling and visualisation. 

In addition, approaches for problems in time will be introduced.

A choice of examples drawn from all areas of the biosciences will allow you to hone your skills on problems of interest before applying them to the experimental design and bioscience techniques strands in this module as well as work in other modules and projects.

Assessment is through statistical and quantitative problem solving approaches related to Experimental Design and Bioscience Techniques.


![](../../pics/58Iend.png)


